flow api tutorial (#161)
jina-ai/examples: flow-api-tutorial/README.md

  <img src="https://github.com/jina-ai/jina/blob/master/.github/badges/twitter-badge.svg"
       alt="tweet button" title="👍Share Jina with your friends on Twitter"></img>
</a>
[![Python 3.7 3.8](https://github.com/jina-ai/jina/blob/master/.github/badges/python-badge.svg "Jina supports Python 3.7 and above")](#)
[![Docker](https://github.com/jina-ai/jina/blob/master/.github/badges/docker-badge.svg "Jina is multi-arch ready, can run on differnt architectures")](https://hub.docker.com/r/jinaai/jina/tags)

</p>

In this demo, we'll use some code snippets to build search system. Make sure you've gone through [Jina 101](https://github.com/jina-ai/jina/tree/master/docs/chapters/101) and understood the Pod, Flow, Yaml before moving on. 

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

> - [Overview](#overview)

Author: Thanks, Nan, I've added the main content and other details you mentioned before. The more I work with you, the more I learn. Thank you for your detailed review and recommendation.

- [Wrap up](#wrap-up)
- [Next Steps](#next-steps)
- [Documentation](#documentation)
- [Community](#community)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
## Overview
Jina supports you to index and search your data in a simple way. And it exposes 3 APIs to each of them, which help you to index and search `ndarray`, `files`, and `lines` data.

### 3 APIs for indexing your data
<p>

1. `index_ndarray()` is the API for indexing `ndarray`. In this example, we created a 3*8 random matrix by numpy. Jina Flow helps to index the matrix. First, we need to import Flow. Then, we add an empty pod named `_logforward` to the Flow. After that, we start the Flow simply by using `index_ndarray()`. After several seconds, you will receive the results in the logs. 
    
    <p>
    

>     In this case, we used several parameters. `_logforward` is an internal yaml file, which locates in `/jina/resources/executors._forward.yml`. In the code `add(uses='_logforward')`, Flow will create an empty pod which just forward structured data to the clients. The structure is defined by `jina.proto`, which is a [protobuf](https://docs.jina.ai/api/jina.proto.html?highlight=proto#module-jina.proto) file. Pods forward the structured data file to each other or to the clients by [gRPC](https://docs.jina.ai/chapters/restapi/index.html?highlight=grpc).

Reviewer: I would suggest to link it here: [protobuf](https://docs.jina.ai/chapters/proto/docs.html)

Author: Yes, you are right. Marvelous!

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Overview](#overview)
- [3 APIs for indexing and searching your data](#3-APIs-for-indexing-and-searching-your-data)
    - [ndarray API](#1-ndarray-api)
    - [files API](#2-files-api)
    - [lines API](#3-lines-api)
- [Wrap up](#wrap-up)
- [Next Steps](#next-steps)
- [Documentation](#documentation)
- [Community](#community)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
## Overview
Jina supports you to index/search your data in a simple way. And it exposes 3 APIs for each of them, which help you to index or search `np.ndarray`, `files`, and `text`.

## 3 APIs for indexing and searching your data

### 1. ndarray API
`index_ndarray()` is the API for indexing `np.ndarray` 

```python
import numpy as np
from jina.flow import Flow
input_data = np.random.random((3,8))
f = Flow().add(uses='_logforward')
with f:
    f.index_ndarray(input_data)
```
    

> * Add an empty pod with config `_logforward` to the Flow. [`_logforward`](https://docs.jina.ai/chapters/simple_exec.html) is a built-in yaml, which just forwards input data to the results, and it locates in `jina/resources/executors._forward.yml`. You can also use your own [yaml](https://docs.jina.ai/chapters/yaml/yaml.html) to organize `pods`.

Reviewer 2: empty -> dummy

Reviewer 2: which just forwards input data to the results,

 which just forwards input message and print it to the log

> * Start the Flow by using API `index_ndarray()`. The results will be printed in the logs. 

Reviewer 2: ```suggestion
* use the Flow to index an `ndarray` by calling the `index_ndarray()` API. 
```

> The data structure for `index_ndarray()` is used to transfer data. As we use `_logforward`, it will be printed out in the logs as follows:

Reviewer 2: ```suggestion
Calling the `index_ndarray()` API will generate requests with the following message.
```

    start_time {
      seconds: 1597931714
      nanos: 347768000
    }
  }
  version {
    jina: "0.4.1"
    proto: "0.0.55"
  }
  num_part: 1
}
request {
  request_id: 1
  index {
    docs {
      id: 1
      weight: 1.0
      length: 100
      blob {
        buffer: "\004@\316\362/D\333?\244>\235\305\027\311\336?\267\210\251\311^\260\345?\366\n(\014\022m\356?\374\262\017\030\036\357\351?-c\300\337\217V\345?\241G\241\352\233\024\356?\340\346lUf\353\350?"
        shape: 8
        dtype: "float64"
      }
    }
    docs {
      id: 2
      weight: 1.0
      length: 100
      blob {
        buffer: "\312Wm\337\250\217\354?t\212\326\020\261\r\320?\254\262\300u<O\323?\340\210\222$\321\216\314?\310.q,+\347\311?&\316\361\310\252R\331?\214\016\201a\231\262\330?\342\231\262\221\343%\324?"
        shape: 8
        dtype: "float64"
      }
    }
    docs {
      id: 3
      weight: 1.0
      length: 100
      blob {
        buffer: "kT\250\372K%\345?\237\017+u\300\227\353?\3668\256\340\251\227\350?\327\006$\032$\002\341?\274\300\3573\371\262\343?\346\371\265dV\330\342?\370\210\360\002P3\340?\022i-\016\374\320\331?"
        shape: 8
        dtype: "float64"
      }
    }
  }
}
```

> The data structure is defined in `jina.proto`. `jina.proto` is a [protobuf](https://docs.jina.ai/chapters/proto/docs.html) file. Pods forward the structured input data to each other or to the clients by [gRPC](https://docs.jina.ai/chapters/restapi/index.html?highlight=grpc) services.

Reviewer 2: ```suggestion
The structure of the message is defined in the format of [protobuf](https://docs.jina.ai/chapters/proto/docs.html). Check more details of the data structure at [`jina.proto` add url here]().  Messages are passed between the Pods in the Flow.
```

> It resolves the issue: [#134](https://github.com/jina-ai/examples/issues/134)

Reviewer: should this be deleted?

Author: Yes, sure

# Overview
There are 3 ways to use index and search method in Jina, we have examples like faiss, southpark and flower-search. This note will help you understand the pattern beyond those 3 examples.

#3 ways to initiate index flow
<p>
 
1.index_ndarray supports ndarray, and you can use this method to index your array data, like pics and sounds. Here is the example:[Faiss Search](https://github.com/jina-ai/examples/tree/master/faiss-search)

```python
import numpy as np
from jina.flow import Flow
flow = Flow().load_config('flow-index.yml')
input_data = np.random.random(10,100)
with flow.build() as fl:
    fl.index_ndarray(input_data)
```
2.index_files supports you to load files. In our example we use jpg files.Here is the example:[flower Search](https://github.com/jina-ai/examples/tree/master/flower-search)
```python
from jina.flow import Flow
f = Flow().load_config('flow-index.yml')
data_path="/your/data/path"
with f:

>     f.index_files(f'{data_path}/*.jpg')

Reviewer 2: ```suggestion
    f.index_files(f'../pokedex-with-bit/pods/*.yml')
```
